### PR TITLE
Create "Fancy Fetch" for niscope

### PR DESCRIPTION
**Pull request by texasaggie97**
- [X] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nimi-python/blob/master/CONTRIBUTING.md).
- [X] I've updated [CHANGELOG.md](https://github.com/ni/nimi-python/blob/master/CHANGELOG.md) if applicable.
- [X] I've added tests applicable for this pull request
### What does this Pull Request accomplish?
* Change `fetch()` to be more user friendly
  * `fetch(timeout=5.0, num_samples=None, fetch_relative_to=None, fetch_offet=None, fetch_record_number=None, fetch_num_records=None, wfm=None)`
  * `fetch_relative_to`, `fetch_offet`, `fetch_record_number`, & `fetch_num_records`: If not `None` will set the corresponding attribute
  * If wfm is not `None` calls the appropriate `fetch_into()` function
  * In Python 3, a memoryview of each record is created and added to the corresponding `wfm_info`
  * In Python 2, an array.array is created and the points are copied.  `memoryview` in Python 2 only supports string and bytearray, not array.array or numpy.ndarray of arbitrary types. :(
* Additional documentation requires #722 
* Instead of returning all of the captured data in a single 1D array that the user has to then correlate different parts to each record, the waveform for a given record is put into the corresponding wfm_info
  before:
  ``` python
  # Assume num records = 2
  wfm, wfm_info = session.channel[[0,1]].fetch(num_samples=1000)
  # Channel 0, record 0
  # wfm_info[0].<absoluteX, etc.> and wfm[0:999]
  # Channel 1, record 0
  # wfm_info[1].<absoluteX, etc.> and wfm[1000:1999]
  # Channel 0, record 1
  # wfm_info[2].<absoluteX, etc.> and wfm[2000:2999]
  # Channel 1, record 1
  # wfm_info[3].<absoluteX, etc.> and wfm[3000:3999]
  ```
  after:
  ``` python
  wfm_info = session.channel[[0,1]].fetch(num_samples=1000, fetch_num_records=2)
  # Channel 0, record 0
  # wfm_info[0].<absoluteX, etc.> and wfm_info[0].wfm
  # Channel 1, record 0
  # wfm_info[1].<absoluteX, etc.> and wfm_info[1].wfm
  # Channel 0, record 1
  # wfm_info[2].<absoluteX, etc.> and wfm_info[1].wfm
  # Channel 1, record 1
  # wfm_info[3].<absoluteX, etc.> and wfm_info[1].wfm
  # Basically, the waveform becomes just one more attribute on the waveform info object
  ```

### What testing has been done?
* Unit
* System

### CODE REVIEW

File: src/niscope/system_tests/test_system_niscope.py [hunk 1]
```diff
@@ -66,6 +67,23 @@ def test_fetch_binary8_into(session):
...
+    test_record_length = 2000
+    test_channels = range(2)
+    test_num_channels = 2
+    wfm = numpy.ndarray(test_num_channels * test_record_length, dtype=numpy.int8)
+    # Initialize with NaN so we can later verify all samples were overwritten by the driver.
+    wfm.fill(float('nan'))
+    session.configure_vertical(test_voltage, niscope.VerticalCoupling.AC)
+    session.configure_horizontal_timing(50000000, test_record_length, 50.0, 1, True)
+    with session.initiate():
+        wfm_infos = session.channels[test_channels].fetch(timeout=5.0, wfm=wfm)
```
Comment (epage) [inline src/niscope/system_tests/test_system_niscope.py]: > If wfm is not None calls the appropriate fetch_into() function

I thought we had decided against doing that kind of stuff ie having functions that both mutate their parameter and return it?

Comment (texasaggie97) [reply src/niscope/system_tests/test_system_niscope.py]: It doesn't directly return the wfm. It puts slices/memoryviews of it into the wfm_info object.

**Comment by texasaggie97** [comment]
I updated the PR description with before and after.
The changelog lists the change. Until #722 is merged, I can't add driver specific, hand maintained documentation.

**Comment by epage** [comment]
> The changelog lists the change. Until #722 is merged, I can't add driver specific, hand maintained documentation.

What do you mean by "the changelog lists the change" and why is driver-specific, hand maintained documentation relevant?

**Comment by epage** [comment]
> In Python 3, a memoryview of each record is created and added to the corresponding wfm_info
> In Python 2, an array.array is created and the points are copied. :(

I assume we talked about this but can't remember the results.  Why is [`memoryview`](https://docs.python.org/2/library/stdtypes.html#memoryview) not used in Python2?

**Comment by texasaggie97** [comment]
> I assume we talked about this but can't remember the results. Why is memoryview not used in Python2?

`memoryview` in Python 2 only supports string and bytearray, not array.array or numpy.ndarray of arbitrary types.

I updated the PR description and added a comment to the source about why.

File: docs/niscope/functions.rst [hunk 2]
```diff
@@ -1212,9 +1212,18 @@ niscope.Session methods
...
         .. code:: python
 
-            session.channels['0,1'].fetch(num_samples, timeout='datetime.timedelta(seconds=5.0)')
+            session.channels['0,1'].fetch(timeout='datetime.timedelta(seconds=5.0)', num_samples=None, fetch_relative_to=None, fetch_offet=None, fetch_record_number=None, fetch_num_records=None, wfm=None)
 
 
+    :param timeout:
+
+
+        The time to wait in seconds for data to be acquired; using 0 for this parameter tells NI-SCOPE to fetch whatever is currently available. Using -1 for this parameter implies infinite timeout.
```
Comment (marcoskirsch) [inline docs/niscope/functions.rst]: if you have control over this text, remove "in seconds" (in general make sure it works for timedelta)

File: docs/niscope/functions.rst [hunk 3]
```diff
@@ -1224,29 +1233,54 @@ niscope.Session methods
...
+        
 
 
-            Returns an array whose length is the **numSamples** times number of waveforms. Call :py:meth:`niscope.Session._actual_num_wfms` to determine the number of waveforms.
+    :type fetch_offet: int
+    :param fetch_record_number:
 
-            
 
+        Zero-based index of the first record to fetch.  Use fetch__num_records to set the number of records to fetch. If not set, use value of NISCOPE_ATTR_RECORD_NUMBER
```
Comment (marcoskirsch) [inline docs/niscope/functions.rst]: two underscores (type)
If you call out the attribute, use its python name. Applies elsewhere.
"If not set, use value of record_number property"

File: generated/niscope/session.py [hunk 4]
```diff
@@ -1839,6 +1839,127 @@ def configure_vertical(self, range, coupling, offset=0.0, probe_attenuation=1.0,
...
+
+            fetch_record_number (int): Zero-based index of the first record to fetch.  Use fetch__num_records to set the number of records to fetch. If not set, use value of NISCOPE_ATTR_RECORD_NUMBER
+
+            fetch_num_records (int): Number of records to fetch. Use -1 to fetch all configured records. If not set, use value of NISCOPE_ATTR_NUM_RECORDS
+
+            wfm (list of float): Returns an array whose length is the **numSamples** times number of waveforms. Call _actual_num_wfms to determine the number of waveforms.
+
+
+        Returns:
+            wfm_info (list of WaveformInfo): Returns an array of classed with the following timing and scaling information about each waveform:
```
Comment (marcoskirsch) [inline generated/niscope/session.py]: In the documentation for this thing there is no `wfm` member.

File: generated/niscope/session.py [hunk 5]
```diff
@@ -1839,6 +1839,127 @@ def configure_vertical(self, range, coupling, offset=0.0, probe_attenuation=1.0,
...
+        num_wfms = self._actual_num_wfms()
+
+        if num_samples is None:
+            num_samples_to_use = int(len(wfm) / num_wfms)
+        else:
+            num_samples_to_use = num_samples
+
+        total_wfm_size = num_samples_to_use * num_wfms
+
+        # If the waveform is not passed in, we will use the builtin python array with element type 'double'
```
Comment (marcoskirsch) [inline generated/niscope/session.py]: I don't think wfm should *ever* be passed in to fancy fetch.

File: generated/niscope/session.py [hunk 6]
```diff
@@ -1839,6 +1839,127 @@ def configure_vertical(self, range, coupling, offset=0.0, probe_attenuation=1.0,
...
+        else:
+            if len(wfm) < total_wfm_size:
+                raise TypeError('The size of the input array is too small for the acquisition. Needs to be {0}, was {1}'.format(total_wfm_size, len(wfm_to_use)))
+
+            wfm_to_use = wfm
+            wfm_info = self.fetch_into(wfm_to_use, timeout)
+
+        # memoryview in Python 2 doesn't support numeric types, so we copy into an array.array to put in the wfm. :( You should be using Python 3!
+        # Or use the _into version. memoryview in Python 2 only supports string and bytearray, not array.array or numpy.ndarray of arbitrary types.
+        if sys.version_info.major < 3:
```
Comment (marcoskirsch) [inline generated/niscope/session.py]: can six help in any way for this?

File: generated/niscope/session.py [hunk 7]
```diff
@@ -1839,6 +1839,127 @@ def configure_vertical(self, range, coupling, offset=0.0, probe_attenuation=1.0,
...
+                        numpy.dtype('int16'): 'h',
+                        numpy.dtype('int32'): 'l',
+                        numpy.dtype('float64'): 'd',
+                    }[wfm_to_use.dtype]
+
+                start = i * num_samples_to_use
+                end = start + num_samples_to_use
+                wfm_info[i].wfm = array.array(typecode, wfm_to_use[start:end])
+        else:
+            # In Python 3 we can use memoryview objects to give us pieces of the underlying array. This is much faster
```
Comment (marcoskirsch) [inline generated/niscope/session.py]: "Python 3 and newer"

File: generated/niscope/session.py [hunk 8]
```diff
@@ -1839,6 +1839,127 @@ def configure_vertical(self, range, coupling, offset=0.0, probe_attenuation=1.0,
...
+        if wfm is None:
+            wfm_to_use, wfm_info = self._fetch(num_samples, timeout)
+        else:
+            if len(wfm) < total_wfm_size:
+                raise TypeError('The size of the input array is too small for the acquisition. Needs to be {0}, was {1}'.format(total_wfm_size, len(wfm_to_use)))
+
+            wfm_to_use = wfm
+            wfm_info = self.fetch_into(wfm_to_use, timeout)
+
+        # memoryview in Python 2 doesn't support numeric types, so we copy into an array.array to put in the wfm. :( You should be using Python 3!
```
Comment (marcoskirsch) [inline generated/niscope/session.py]: move the comment inside the if statement since it only applies to the `sys.version_info.major < 3` case.

File: generated/niscope/session.py [hunk 9]
```diff
@@ -1839,6 +1839,127 @@ def configure_vertical(self, range, coupling, offset=0.0, probe_attenuation=1.0,
...
+        import sys
+
+        if num_samples is None and wfm is None:
+            raise TypeError("Either 'num_samples' or 'wfm' must be set")
+
+        # Check optional parameters
+        if fetch_relative_to is not None:
+            self.fetch_relative_to = fetch_relative_to
+        if fetch_offet is not None:
+            self.fetch_offet = fetch_offet
```
Comment (marcoskirsch) [inline generated/niscope/session.py]: This is a hidden side effect. It won't be obvious that passing the parameter will change the state of the session.

Ideally, I would get rid of the (public) attributes so that there's only one way to do this. But that cannot be done because of _into flavor. Or can it? We'd have to add them as parameters as well. Then maybe they don't need to all be optional parameters.

Alternatively, the correct solution is for the parameters to not have a visible effect on the Session state. How? I think we'd have to use some sort of property context manager that on __exit__ restores the old value.

What do you think?

Comment (marcoskirsch) [reply generated/niscope/session.py]: Also, I don't see where this gets used later on. Do these not have an effect on how you interpret the waveform data?

File: generated/niscope/unit_tests/mock_helper.py [hunk 10]
```diff
@@ -59,6 +59,9 @@ def __init__(self):
         self._defaults['Disable']['return'] = 0
         self._defaults['ExportSignal'] = {}
         self._defaults['ExportSignal']['return'] = 0
+        self._defaults['FancyFetch'] = {}
```
Comment (marcoskirsch) [inline generated/niscope/unit_tests/mock_helper.py]: #695 keeps nagging at me.

File: src/niscope/examples/niscope_fetch.py [hunk 11]
```diff
@@ -13,9 +13,9 @@ def example(resource_name, channels, options, length, voltage):
         session.configure_vertical(range=voltage, coupling=niscope.VerticalCoupling.AC)
         session.configure_horizontal_timing(min_sample_rate=50000000, min_num_pts=length, ref_position=50.0, num_records=1, enforce_realtime=True)
         with session.initiate():
-            wfm, wfm_infos = session.channels[channels].fetch(num_samples=length)
-        print('Number of samples acquired: {:,}\n'.format(len(wfm)))
+            wfm_infos = session.channels[channels].fetch(num_samples=length)
```
Comment (marcoskirsch) [inline src/niscope/examples/niscope_fetch.py]: How about `wfm_infos` -> `waveforms` ?

File: src/niscope/metadata/functions_addon.py [hunk 12]
```diff
@@ -270,14 +272,52 @@
...
             },
             {
-                'direction': 'out',
+                'direction': 'in',
+                'enum': 'FetchRelativeTo',
+                'default_value': None,
+                'name': 'fetchRelativeTo',
+                'type': 'ViInt32',
+                'documentation': {
+                    'description': 'Position to start fetching within one record. If not set, use value of NISCOPE_ATTR_FETCH_RELATIVE_TO',
```
Comment (marcoskirsch) [inline src/niscope/metadata/functions_addon.py]: NISCOPE_ATTR_FETCH_RELATIVE_TO didn't get converted to python_name by the generator.

File: src/niscope/system_tests/test_system_niscope.py [hunk 13]
```diff
@@ -66,6 +67,23 @@ def test_fetch_binary8_into(session):
...
+    test_record_length = 2000
+    test_channels = range(2)
+    test_num_channels = 2
+    wfm = numpy.ndarray(test_num_channels * test_record_length, dtype=numpy.int8)
+    # Initialize with NaN so we can later verify all samples were overwritten by the driver.
+    wfm.fill(float('nan'))
+    session.configure_vertical(test_voltage, niscope.VerticalCoupling.AC)
+    session.configure_horizontal_timing(50000000, test_record_length, 50.0, 1, True)
+    with session.initiate():
+        wfm_infos = session.channels[test_channels].fetch(wfm=wfm)
```
Comment (marcoskirsch) [inline src/niscope/system_tests/test_system_niscope.py]: Interesting.
You're using type of wfm to determine which binary flavor to call into? I was under the impression fancy fetch was ease of use so it should only support float. In the name of KISS I would remove this flexibility. If you think we should keep it, I'd like to use a different way of selecting which underlying thing to call because it's odd to modify the input param _and_ also return it (or memory views to it?).

Client wants fast? Call the appropriate fetch_into. Or use C, really.

File: src/niscope/templates/session.py/fancy_fetch.py.mako [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+%>\
+    def ${f['python_name']}${suffix}(${helper.get_params_snippet(f, helper.ParameterUsageOptions.SESSION_METHOD_DECLARATION)}):
+        '''${f['python_name']}
+
+        ${helper.get_function_docstring(f, False, config, indent=8)}
+        '''
+        import sys
+
+        if num_samples is None and wfm is None:
+            raise TypeError("Either 'num_samples' or 'wfm' must be set")
```
Comment (marcoskirsch) [inline src/niscope/templates/session.py/fancy_fetch.py.mako]: I vote: remove wfm parameter, make num_samples not optional.

File: src/niscope/templates/session.py/fetch_waveform.py.mako [hunk 15]
```diff
@@ -15,13 +15,13 @@
         num_samples = int(len(wfm) / self._actual_num_wfms())
 
         if wfm.dtype == numpy.float64:
-            return self._fetch_into(num_samples=num_samples, wfm=wfm, timeout=timeout)
+            return self._fetch_into_numpy(num_samples=num_samples, wfm=wfm, timeout=timeout)
         elif wfm.dtype == numpy.int8:
-            return self._fetch_binary8_into(num_samples=num_samples, wfm=wfm, timeout=timeout)
+            return self._fetch_binary8_into_numpy(num_samples=num_samples, wfm=wfm, timeout=timeout)
```
Comment (marcoskirsch) [inline src/niscope/templates/session.py/fetch_waveform.py.mako]: why rename these?
won't they work with array.array anymore?

File: docs/niscope/functions.rst (hunk 2 above)
Comment (texasaggie97) [reply docs/niscope/functions.rst]: Removed, but added seconds after -1:
'The time to wait for data to be acquired; using 0 for this parameter tells NI-SCOPE to fetch whatever is currently available. Using -1 seconds for this parameter implies infinite timeout.'

File: docs/niscope/functions.rst (hunk 3 above)
Comment (texasaggie97) [reply docs/niscope/functions.rst]: Two underscores fixed.

Anything with <DRIVER>_ATTR_<C_NAME> should be replaced with a link to session.<python_name>. I need to figure out why these aren't.

File: generated/niscope/session.py (hunk 4 above)
Comment (texasaggie97) [reply generated/niscope/session.py]: Added with text 'waveform array whose length is the **numSamples**'

File: generated/niscope/session.py (hunk 9 above)
Comment (texasaggie97) [reply generated/niscope/session.py]: These attributes control how fetch works. I don't think it is a problem that we are setting the attributes on the session since the purpose of this is to simplify the fetch.

Even if we added them to the _into dispatcher function, they would still be optional, it's just they would always be set since the default value wouldn't be `None`, but whatever the default value is, if there is a single value that works. There may not be if there are device specific defaults.

These do not control how I interpret the waveform data, they are used by the driver to determine exactly which samples to fetch.

File: generated/niscope/session.py (hunk 5 above)
Comment (texasaggie97) [reply generated/niscope/session.py]: Why? You don't want to support ease of use with the advantages of numpy?

File: generated/niscope/session.py (hunk 8 above)
Comment (texasaggie97) [reply generated/niscope/session.py]: Moved

File: generated/niscope/session.py (hunk 6 above)
Comment (texasaggie97) [reply generated/niscope/session.py]: I don't see anything in `six` for `memoryview`'s

File: generated/niscope/session.py (hunk 7 above)
Comment (texasaggie97) [reply generated/niscope/session.py]: Changed, even though we don't know what Python 4 will bring.

File: generated/niscope/unit_tests/mock_helper.py (hunk 10 above)
Comment (texasaggie97) [reply generated/niscope/unit_tests/mock_helper.py]: Yeah, this'll get fixed eventually, but since it won't affect the API, it has a lower priority.

File: src/niscope/examples/niscope_fetch.py (hunk 11 above)
Comment (texasaggie97) [reply src/niscope/examples/niscope_fetch.py]: Changed

File: src/niscope/system_tests/test_system_niscope.py (hunk 13 above)
Comment (texasaggie97) [reply src/niscope/system_tests/test_system_niscope.py]: If you have to call _into, you don't get the ease of use benefits. 

We don't really return the same parameter that was passed in. We return either new `array.array`'s or `memoryview`'s of the array passed in, but these correspond to the acquired records and are fairly transformed from what was originally passed in.

It seems like the benefits of the fancy fetch should be available along with the speed of numpy and _into.

File: src/niscope/templates/session.py/fancy_fetch.py.mako (hunk 14 above)
Comment (texasaggie97) [reply src/niscope/templates/session.py/fancy_fetch.py.mako]: I vote: keep wfm parameter, num_samples optional.

File: src/niscope/templates/session.py/fetch_waveform.py.mako (hunk 15 above)
Comment (texasaggie97) [reply src/niscope/templates/session.py/fetch_waveform.py.mako]: They don't work with `array.array` now.

I renamed these because I wanted to add `array.array` but saw it was going to be more involved that I wanted for this PR. I decided to leave the '_numpy' suffix because I would like to get '_into_array' supported but in a different PR.

File: src/niscope/metadata/functions_addon.py (hunk 12 above)
Comment (texasaggie97) [reply src/niscope/metadata/functions_addon.py]: Yeah, I need to figure out why. I'll keep looking into it.

File: src/niscope/system_tests/test_system_niscope.py (hunk 13 above)
Comment (marcoskirsch) [reply src/niscope/system_tests/test_system_niscope.py]: "fancy fetch" is also "easy to use fetch"

I expect +90% of users to want scaled float results. They will use "fancy fetch".
I expect only a very small share of users to want to extract absolute maximum performance and don't need things scaled for some rare reason. They will use something other than float and the "fetch_into" methods.
Users who want "easy to use fetch" _and_ unscaled data will be extremely rare if any. I don't think adding complexity to "fancy fetch"'s usage that impact +90% of the users is the right tradeoff.

File: generated/niscope/session.py (hunk 5 above)
Comment (marcoskirsch) [reply generated/niscope/session.py]: Because it makes the API more complex for the vast majority of users.

File: generated/niscope/session.py (hunk 9 above)
Comment (marcoskirsch) [reply generated/niscope/session.py]: Ok, they'd still be optional.

But the root issue remains, that there is a hidden side effect / WTH factor. A reasonable user wouldn't expect that passing an attribute value would change the state of the Session object.

We can live with it, but I think it's unexpected and thus an issue.

Comment (texasaggie97) [reply generated/niscope/session.py]: I think we are beyond where I have a strong opinion on which way to go with this. I will defer to you or @auchter.

I do agree if we make them private that that should be in a separate PR.

File: generated/niscope/session.py (hunk 5 above)
Comment (texasaggie97) [reply generated/niscope/session.py]: The vast majority of users won't deal with it and will leave it with the default of `None`.

File: src/niscope/system_tests/test_system_niscope.py (hunk 13 above)
Comment (texasaggie97) [reply src/niscope/system_tests/test_system_niscope.py]: I don't think it add much complexity at all. Not at the API level (one additional, optional parameter) or in the implementation. 
I do think the additional benefits of encapsulating the property calls as well as bundling the actual waveform with the rest of the waveform information for each individual record is worth it.

File: generated/niscope/session.py (hunk 5 above)
Comment (marcoskirsch) [reply generated/niscope/session.py]: > The vast majority of users won't deal with it and will leave it with the default of None.

They will have to understand that the right thing to do for them is to rely on the default. So they are dealing with it.

File: src/niscope/metadata/functions_addon.py (hunk 12 above)
Comment (texasaggie97) [reply src/niscope/metadata/functions_addon.py]: See #749

File: generated/niscope/session.py (hunk 9 above)
Comment (auchter) [reply generated/niscope/session.py]: I agree with Marcos: I don't like that hidden side effect either.

I lean pretty strongly towards deprecating the fetch attributes in favor of making them parameters. So I think what you have here is a good start, but instead of having these parameters be optional and None, they should be optional and specified to the same defaults that the driver uses. So probably like `def fetch(self, timeout=5.0, samples=1000, records=1, relative_to=enums.FetchRelativeTo.PRETRIGGER, offset=0)`.

I think the goal should be that it's possible to look at a call to fetch and understand, in isolation, what data it's going to retrieve.

Also, looking at that signature... why is `channels` not a parameter to fetch?

Comment (marcoskirsch) [reply generated/niscope/session.py]: > ...but instead of having these parameters be optional and None, they should be optional and specified to the same defaults that the driver uses.

What worries me about this is that there is an assumption that the driver will use the same defaults for all devices. This isn't guaranteed, is it? Or perhaps if the assumption is broken and the API has different defaults, it's not a huge deal?

> Also, looking at that signature... why is channels not a parameter to fetch?

Check [the green tip](http://nimi-python.readthedocs.io/en/0.7.0/niscope/functions.html#niscope.Session.fetch). :)
Which I just noticed could use some updating in the text.

Comment (marcoskirsch) [reply generated/niscope/session.py]: @texasaggie97 I think, summarizing @auchter's suggestion:

1. All fetch methods should take the parameters.
2. We should make the properties private. This is not something the generator supports yet so it'd be a different PR.
3. The parameters should default to explicit values rather than None for discoverability.
4. The defaults should match driver defaults and hope those never change 🤞

I like that this is an overall simplification of the usage API. And it eliminates the side effect because it's no longer visible to clients. I believe all public fetch methods have some hand-coding so it should not be too bad.

**Comment by texasaggie97** [comment]
@auchter 

> Definitely misspoke (typed?) yesterday when I suggested 1000 for num samples. I think this should default to -1, to fetch the entire configured record.

`num_samples` now has a default of `None` (not -1) since it is more Pythonic. I need the actual number so if it is `None`, I go query the actual number in order to allocate the array.

**Comment by texasaggie97** [comment]
> I'd also suggest removing the "fetch_" prefix from these params since it's a but redundant.

Does this include `fetch_offset`? Would just `offset` be too generic?

**Comment by texasaggie97** [comment]
Made fetch attributes private and added as parameters to `fetch_into()`.

**Comment by marcoskirsch** [comment]
Changelog needs updating for the now hidden attributes.

File: docs/niscope/functions.rst [hunk 16]
```diff
@@ -1212,9 +1212,18 @@ niscope.Session methods
 
         .. code:: python
 
-            session.channels['0,1'].fetch(num_samples, timeout='datetime.timedelta(seconds=5.0)')
+            session.channels['0,1'].fetch(timeout='datetime.timedelta(seconds=5.0)', num_samples=None, relative_to=niscope.FetchRelativeTo.PRETRIGGER, offset=0, record_number=0, num_records=-1)
```
Comment (marcoskirsch) [inline docs/niscope/functions.rst]: Since num_samples=None is the default and means "you choose for me", why not make num_records the same?

File: docs/niscope/session.rst [hunk 17]
```diff
@@ -108,13 +108,13 @@ niscope.Session
    +-----------------------------------------------------+---------------------------------------+
    | :py:attr:`fetch_meas_num_samples`                   | int                                   |
    +-----------------------------------------------------+---------------------------------------+
-   | :py:attr:`fetch_num_records`                        | int                                   |
+   | :py:attr:`_fetch_num_records`                       | int                                   |
```
Comment (marcoskirsch) [inline docs/niscope/session.rst]: Private attributes are still in the RST, they should be removed.

File: CHANGELOG.md [hunk 18]
```diff
@@ -39,6 +39,7 @@ All notable changes to this project will be documented in this file.
 * ### NI-SCOPE
     * #### Added
     * #### Changed
+        * `niscope.fetch()` now takes additional parameters for associated attributes and add resulting record as part of the waveform info
```
[the remark this answers is not in the document]
Comment (texasaggie97) [reply CHANGELOG.md]: Added

File: CHANGELOG.md [hunk 19]
```diff
@@ -111,7 +112,7 @@ All notable changes to this project will be documented in this file.
         * Repeated capablilites - See [#737](https://github.com/ni/nimi-python/issues/737) for discussion:
             * `channel` repeated capability
     * #### Changed
-        * `niscope.Session()` no longer takes id_query parameter (Fix [#670](https://github.com/ni/nimi-python/issues/670))
+        * `niscope.Session()` no longer takes id_query parameter
```
[the remark this answers is not in the document]
Comment (texasaggie97) [reply CHANGELOG.md]: Fixed

File: docs/niscope/functions.rst (hunk 16 above)
Comment (texasaggie97) [reply docs/niscope/functions.rst]: The difference is that `None` means go to the driver and get a number that I use to allocate memory, and the -1 is passes into the driver as is.

It would make the logic look like this:
```python
self._fetch_num_records = -1 if num_records is None else num_records
if num_samples is None:
    num_samples = self.horz_record_length
```

Is that okay?

File: docs/niscope/session.rst (hunk 17 above)
Comment (texasaggie97) [reply docs/niscope/session.rst]: Removed private attributes.

File: docs/niscope/functions.rst (hunk 16 above)
Comment (marcoskirsch) [reply docs/niscope/functions.rst]: I think that's ok. The important thing is the interface.

File: CHANGELOG.md [hunk 20]
```diff
@@ -39,6 +39,12 @@ All notable changes to this project will be documented in this file.
 * ### NI-SCOPE
     * #### Added
     * #### Changed
+        * `niscope.fetch()` and `niscope.fetch_into()` now takes additional parameters for associated attributes and add resulting record as part of the waveform info
+        * Following properties are now private (added as parameters to fetch calls):
```
Comment (marcoskirsch) [inline CHANGELOG.md]: Minor: The fact that they are private is an implementation detail. From the API point of view, the properties were removed.

`Following properties are removed. Use new parameters to `Session.fetch()` and `Session.fetch_into()`

File: CHANGELOG.md [hunk 21]
```diff
@@ -39,6 +39,12 @@ All notable changes to this project will be documented in this file.
 * ### NI-SCOPE
     * #### Added
     * #### Changed
+        * `niscope.fetch()` and `niscope.fetch_into()` now takes additional parameters for associated attributes and add resulting record as part of the waveform info
```
Comment (marcoskirsch) [inline CHANGELOG.md]: Should be `niscope.Session.fetch()` or since you are inside the NI-SCOPE section `Session.fetch()`

File: docs/niscope/functions.rst [hunk 22]
```diff
@@ -1191,7 +1191,7 @@ niscope.Session methods
 
     :type signal_identifier: str
 
-.. py:method:: fetch(num_samples, timeout='datetime.timedelta(seconds=5.0)')
+.. py:method:: fetch(timeout='datetime.timedelta(seconds=5.0)', num_samples=None, relative_to=niscope.FetchRelativeTo.PRETRIGGER, offset=0, record_number=0, num_records=-1)
```
Comment (marcoskirsch) [inline docs/niscope/functions.rst]: I think timeout is (almost?) always the last parameter. Let's do that here too.

File: docs/niscope/functions.rst (hunk 16 above)
Comment (marcoskirsch) [reply docs/niscope/functions.rst]: On one hand, I think public API is what matters. On the other hand, we have fully generated functions in which -1 means "automatic". I don't have a strong opinion either way, as long as docstring is helpful.

File: generated/niscope/session.py [hunk 23]
```diff
@@ -1899,6 +1914,97 @@ def configure_vertical(self, range, coupling, offset=0.0, probe_attenuation=1.0,
         errors.handle_error(self, error_code, ignore_warnings=False, is_error_handling=False)
         return
 
+    def fetch(self, timeout=datetime.timedelta(seconds=5.0), num_samples=None, relative_to=enums.FetchRelativeTo.PRETRIGGER, offset=0, record_number=0, num_records=-1):
+        '''fetch
+
+        Returns the waveform from a previously initiated acquisition that the
+                        digitizer acquires for the specified channel. This method returns
```
Comment (marcoskirsch) [inline generated/niscope/session.py]: indentation is odd here. How does this affect things?

File: src/niscope/examples/niscope_fetch.py [hunk 24]
```diff
@@ -13,11 +13,11 @@ def example(resource_name, channels, options, length, voltage):
         session.configure_vertical(range=voltage, coupling=niscope.VerticalCoupling.AC)
         session.configure_horizontal_timing(min_sample_rate=50000000, min_num_pts=length, ref_position=50.0, num_records=1, enforce_realtime=True)
         with session.initiate():
-            wfm, wfm_infos = session.channels[channels].fetch(num_samples=length)
-        print('Number of samples acquired: {:,}\n'.format(len(wfm)))
-        for i in range(len(wfm_infos)):
+            waveforms = session.channels[channels].fetch(num_samples=length)
+        for i in range(len(waveforms)):
+            print('Number of samples acquired: {:,}\n'.format(len(waveforms[i].wfm)))
```
Comment (marcoskirsch) [inline src/niscope/examples/niscope_fetch.py]: I think this line needs to go after the header printed in line 19. Right?

File: src/niscope/metadata/functions_addon.py [hunk 25]
```diff
@@ -425,29 +495,29 @@
...
-    'FetchBinary32':                                 { 'method_templates': [
-        { 'session_filename': 'numpy_read_method', 'method_python_name_suffix': '_into', },
+    'FetchBinary32':   { 'method_templates': [
+        { 'session_filename': 'numpy_read_method', 'method_python_name_suffix': '_into_numpy', },
     ], },
-    'Fetch':                { 'method_templates': [
+    'Fetch':           { 'method_templates': [
         { 'session_filename': 'default_method', 'method_python_name_suffix': '', },
-        { 'session_filename': 'numpy_read_method', 'method_python_name_suffix': '_into', },
+        { 'session_filename': 'numpy_read_method', 'method_python_name_suffix': '_into_numpy', },
```
Comment (marcoskirsch) [inline src/niscope/metadata/functions_addon.py]: The reason these were renamed is because they don't work as is with array.array, right?
And the plan is to fix that and change the name back? If so, I don't think the rename is necessary, just an issue.

File: src/niscope/templates/session.py/fancy_fetch.py.mako [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+%>\
+    def ${f['python_name']}${suffix}(${helper.get_params_snippet(f, helper.ParameterUsageOptions.SESSION_METHOD_DECLARATION)}):
+        '''${f['python_name']}
+
+        ${helper.get_function_docstring(f, False, config, indent=8)}
+        '''
+        import sys
+
+        # Set the fetch attributes
+        with _NoChannel(session=self):
```
Comment (marcoskirsch) [inline src/niscope/templates/session.py/fancy_fetch.py.mako]: I think this looks very good, very self-documented.

File: src/niscope/templates/session.py/fancy_fetch.py.mako [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+                end = start + num_samples
+                wfm_info[i].wfm = array.array('d', wfm[start:end])
+        else:
+            # In Python 3 and newer we can use memoryview objects to give us pieces of the underlying array. This is much faster
+            mv = memoryview(wfm)
+
+            for i in range(len(wfm_info)):
+                start = i * num_samples
+                end = start + num_samples
+                wfm_info[i].wfm = mv[start:end]
```
Comment (marcoskirsch) [inline src/niscope/templates/session.py/fancy_fetch.py.mako]: It seems most of the code in the the if / else blocks is the same except for this line and the creation of the memory view. If you think you can reduce repetition without making it dirtier, that would be preferable.

File: src/niscope/templates/session.py/fancy_fetch.py.mako [hunk 28]
```diff
@@ -0,0 +1,43 @@
+<%page args="f, config, method_template"/>\
+<%
+    '''Forwards to _fetch_double()'''
```
Comment (marcoskirsch) [inline src/niscope/templates/session.py/fancy_fetch.py.mako]: Is this still accurate?

File: CHANGELOG.md (hunk 20 above)
Comment (texasaggie97) [reply CHANGELOG.md]: Changed

File: CHANGELOG.md (hunk 21 above)
Comment (texasaggie97) [reply CHANGELOG.md]: Changed

File: generated/niscope/session.py (hunk 23 above)
Comment (texasaggie97) [reply generated/niscope/session.py]: It doesn't. It is cosmetic, but now fixed.

File: src/niscope/templates/session.py/fancy_fetch.py.mako (hunk 28 above)
Comment (texasaggie97) [reply src/niscope/templates/session.py/fancy_fetch.py.mako]: Actually, forwards to _fetch(). Fixed

**Comment by marcoskirsch** [comment]
Note flake8 failed.
